### DISCUSSION TITLE
Pattern matching

### DISCUSSION THREAD

**Discussion by gustavopch** [DanielXMoore/Civet#150] [Ideas]
So, basically implement https://github.com/tc39/proposal-pattern-matching.

Note: the proposal has an `is` keyword that's incompatible with Civet's `is`: https://github.com/tc39/proposal-pattern-matching#is-expression. As it's just a proposal in stage 1 and may still change a lot, Civet may keep `is` as it is now and use a different keyword like `matches` instead. I mean, maybe Civet becomes popular enough in the mean time so to inform the proposal as it progresses to stage 4, which is likely to take several years, if it ever happens.

There's an interesting pattern matching library for TypeScript: https://github.com/gvergnaud/ts-pattern. It may inform the implementation in Civet. Perhaps even talk to the maintainers and see if they'd like to contribute to Civet?

ReScript also has pattern matching, so analyzing the JS generated by it may give some insights: https://rescript-lang.org/docs/manual/latest/pattern-matching-destructuring.

I believe pattern matching is the single feature most likely to make FP folks who have to use TS at work (instead of their favorite FP language) be immediately interested in Civet.

**Comment by STRd6** [comment]
Looking at the ReScript transpilation it looks like this should be fairly straightforward to add. We already extend the `switch` statement and we can compile it differently depending on whether it contains `|` or `case/when`.

Some sort of pattern matching will be a great feature for Civet. I'm sure we'll be able to think it through as a community and settle on something really good 🚀

**Reply by gustavopch** [reply]
Thought it would stray a lot from the TC39 proposal, right? Not necessarily a problem. Stage 1 is too premature to care a lot. I mean, at the time it lands, we may already have been replaced by AI anyway. lol

**Reply by STRd6** [reply]
I think the value of the TC39 proposal is in identifying the problem and the goals, much less in any specific syntax. I don't think the proposal will make it into the JS standard in its current form. We also have the advantage of building on top of TypeScript so there are some things we can solve more cleanly than JS alone can.

**Reply by gustavopch** [reply]
Keep in mind that ReScript also uses pattern matching for exception handling:
```rescript
try {
  doSomething()
} catch {
| Error(message) => Js.log(message)
}

switch doSomething() {
| result => Js.log(result)
| exception Error(message) => Js.log(message)
}
```

Also note that in both cases it's performing two things:
- Checking if the caught value is an `Error` variant.
- Extracting the `message` property from it.

In TS, we don't have variants as in ReScript, so checking if a value matches something must be done by:
- Checking the properties (duck typing).
- Using `instanceof`.
- Using a function like `Array.isArray()`.

We have to think if it's actually possible to use ReScript syntax here, without the concept of variants.

On the other hand, if we try to follow TC39's proposal, we can do something along the lines of:
```ts
try
  doSomething()
catch match error
  when { message } if error instanceof Error: log message
```

This approach may lend itself better for TS. I took the catch integration from https://github.com/tc39/proposal-pattern-matching#integration-with-catch.

**Reply by edemaine** [reply]
Oh man, I've wanted pattern matching for caught errors since forever... even Python had a basic version in 1.5.2 😅

**Reply by gustavopch** [reply]
@STRd6 BTW exhaustiveness checking is very important to have ([ts-pattern](https://github.com/gvergnaud/ts-pattern) has it). I believe it's a hard thing to do because now it's not simply a matter of transpiling, but also ensuring the type inference in the transpiled code can detect if all possible cases are being handled or not.

**Comment by gustavopch** [comment]
Exploring the TC39 syntax (or at least something similar to it):
```ts
ok, items := match await doSomething()
  when { ok: true, items } if Array.isArray(items): true, items
  default: false, []
  catch error:
    when _ if error instanceof NotFound: false, []
    when _: do
      log error?.message
      false, null // Implicit return

match ok, items // Matching 2 things at the same time (as supported by ReScript)
  when true, _ if items.length > 10: do
    log 'found many items'
  when true, _: do
    log 'found few items'
  when false, _: do
    log 'failed'
```

The `_` is borrowed from ReScript and means "any value". With that, we don't need a `default` because it's just `when _`. It's important if we want to allow matching against multiple values at the same time.

**Comment by tomByrer** [comment]
From [Grain](https://grain-lang.org/docs/guide/pattern_matching):
```ts
enum PizzaTopping { Cheese, Pepperoni, Peppers, Pineapple }

match (topping) {
  Cheese => print("Would it really be pizza without it?"),
  Pepperoni => print("An instant classic."),
  Peppers => print("For those who like to spice things up."),
  Pineapple => print("You do you.")
}
```
requested to go here [by Daniel](https://discord.com/channels/933472021310996512/933472021310996515/1062960401895149599)
I'm not sure if the thick-arrows would work, but the `match` here looks really clean, fast to scan & understand.

**Comment by STRd6** [comment]
Hypothetical syntax inspired by TS-Pattern semantics:

```typescript
type Data =
  | { type: 'text'; content: string }
  | { type: 'img'; src: string };

type Result =
  | { type: 'ok'; data: Data }
  | { type: 'error'; error: Error };

const result: Result = ...;

switch result
| { type: 'error' } => <p>Oups! An error occured</p>
| { type: 'ok', data: { type: 'text' } } (res) => <p>{res.data.content}</p>
| { type: 'ok', data: { type: 'img', src } } (src) => <img src={src} />
```

LSP will show an error if non-exhaustive. Names in the matching/binding pattern that don't have a value will be bound to the `() =>` parameters left to right in the order they appear. If no bindings are present the default value is the entire result.

Edit: we can probably remove the `() =>` if we have a result reference symbol. Any bindings will be available on the right hand side.

```typescript
switch result
| { type: 'error' } <p>Oups! An error occured</p>
| { type: 'ok', data: { type: 'text' } } <p>{^.data.content}</p>
| { type: 'ok', data: { type: 'img', src } } <img src={src} />
```

Using `^` as placeholder topic reference.

**Comment by STRd6** [comment]
```typescript
switch result
  { type: 'error' }
    <p>Oups! An error occured</p>
  { type: 'ok', data: { type: 'text' } }
    <p>{^.data.content}</p>
  { type: 'ok', data: { type: 'img', src } }
    <img src={src} />
```

Maybe we can drop the `|` too.

**Reply by gustavopch** [reply]
I like the clean syntax. Could we also use `if` expressions and would it work with the exhaustive checking?
```ts
switch error
  { message } if error instanceof NotFoundError
    log `Not found: message`
  ...
```
Must also think about: matching 2 things at once, default clause, catch integration?

BTW `^` conflicts with https://github.com/DanielXMoore/Civet/discussions/113 and maybe it's not even needed?

**Reply by STRd6** [reply]
You're right that it's probably not necessary. It may not conflict since the following characters would be different but I'd be fine moving forward without it.

**Reply by gustavopch** [reply]
Just adding here, in my example
above, when you access `error` inside a clause, it shouldn’t be the same `error` passed to the `switch`, but rather the narrowed version.

But maybe we should also try to find another syntax for the auto-import proposal.

**Reply by STRd6** [reply]
I think if the conditionals are set up properly TypeScript will automatically narrow the type.

**Reply by tomByrer** [reply]
I prefer the `|` is dropped also; reminds me too much of pipes.

But I prefer the placeholder is the same character used by pipes & also in switch.  I'm impartial what char ATM, but just want across the board standard.

**Comment by STRd6** [comment]
We may also be able to do stuff like:

```typescript
switch result
  instanceof NotFoundError and { message }
    log `Not found: ${message}`
  ...
```

**Comment by STRd6** [comment]
More prior art to consider: https://doc.rust-lang.org/book/ch18-01-all-the-places-for-patterns.html
